pkg/arvo/neo/cod/std/src/fil/tasks.md

```diff
@@ -0,0 +1,624 @@
+# Tutorial 4: Tasks
+Let’s take a look at the Tasks shrub. From within the Sky frontend, you can create, edit, reorder, and delete nested tasks and subtasks. Checking off all subtasks will mark the parent as complete, which involves some interaction between parent and child shrubs.
+
+This time there are no new concepts on the backend, but there’s much more going on in the UI than we’ve looked at in previous tutorials.
+
+In this lesson we’ll see how shrubs keep their parents informed of state changes using the `%gift` poke. This is the most complex UI we’ve looked at yet, so we’ll also focus on the `/con` files.
+
+## /imp/task.hoon
+Tasks only needs one `/imp` file: `task.hoon`. The Tasks frontend shows you some tasks that may or may not have other tasks as children.
+
+```hoon
+/@  task  ::  [text=cord done=? order=(list path]
+::
+::  $task-diff
+::  $%  [%new =task prepend=?]
+::      [%edit text=cord done=?]
+::      [%oust =pith]
+::      [%reorder order=(list pith)]
+::  ==
+/@  task-diff
+=>
+::
+::  helper core
+|%
+::
+::  check if all kids are completed
+++  check-kids
+  |=  =bowl:neo
+  ^-  ?
+  ?:  =([~ ~] kids.bowl)
+    %.y
+  =/  piths  ~(tap in ~(key by ~(tar of:neo kids.bowl)))
+  %+  levy  piths
+  |=  =pith
+  =/  =task
+    !<  task
+    q.pail:(need (~(get by ~(tar of:neo kids.bowl)) pith))
+  done.task
+::
+::  assign a unique numerical ID to a new subtask
+++  assign-name
+  |=  =bowl:neo
+  ^-  @ud
+  ?:  =([~ ~] kids.bowl)  1
+  =/  sorted-names=(list @ud)
+    %-  sort  :_  lth
+    %+  turn  ~(tap by ~(tar of:neo kids.bowl))
+      |=  [=pith =idea:neo]
+      +:(,[%ud @ud] (rear pith))
+  =/  last-name=@ud  (rear sorted-names)
+  =/  name-missing=(list @ud)
+    %+  skim  (gulf 1 last-name)
+    |=  n=@ud
+    =(~ (find ~[n] sorted-names))
+  ?~  name-missing  +(last-name)
+  (rear name-missing)
+--
+::
+::  outer core
+^-  kook:neo
+|%
+::
+::  state is a %task
+++  state  pro/%task
+::
+::  we accept %gift and %task-diff pokes
+++  poke   (sy %task-diff %gift ~)
+::
+::  we define one generation of kids at
+::  /path/to/this/task/<@ud> which are
+::  also tasks like their parent, which
+::  means those kids also define one
```

Comment:
This is incorrect; because any implementation of the %task protocol can fulfill this requirement, and not all implementations of the %task protocol will necessarily define kids, we don't know anything about our grandkids here.

...writing this down, I'm realizing that we should change the state of the %task protocol to signify whether a task can be modified, rather than checking its kids for this. @YunPearce #49 

Yun is going to make this change.

Reply:
I made the change! @bonbud-macryg take note.